opt: do some cleanup and re-packaging

andy-kimball opened the pull request:
This PR contains two commits that cleanup the usage and packaging
of the optimizer.

NOTE: From now on, list item expressions need to be constructed using
Factory rather than via direct construction of the struct.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43170)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks for the list items cleanup, very useful! I believe it fixes #36148 as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-:-Lw9kEDC8Nhp8KJyykMr:b-2nopfz)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/check_expr.go, line 268 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-Lw9dP0ZEYFmGwP2CsqU:-Lw9dP0ZEYFmGwP2CsqV:bhy3l9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc18a6fc4bb4fbf118f4fc6c98e5642dc6adae7/pkg/sql/opt/memo/check_expr.go#L268)):*
> ```Go
> func checkFilters(filters FiltersExpr) {
> 	for _, item := range filters {
> 		if opt.IsListItemOp(item.Condition) {
> ```

Out of curiosity, why did this go away? Is this now allowed?
___
*[pkg/sql/opt/optbuilder/srfs.go, line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-Lw9gO831BGo3zaEAhOY:-Lw9gO84DZzZsDvr-ZJM:b5x9p0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/optbuilder/srfs.go#L116)):*
> ```Go
> 		zip[i] = b.factory.ConstructZipItem(
> 			b.buildScalar(texpr, inScope, outScope, outCol, nil),
> 			make(opt.ColList, len(outScope.cols)-startCols),
> ```

Not crazy about this, feels strange to call a `Construct` method with a not-yet populated slice (in principle, it could be using those values to populate the props). I'd first populate the slice then call Construct.
___
*[pkg/sql/opt/optbuilder/srfs.go, line 178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-Lw9gWT5-ZzUSHQs2sgD:-Lw9gWT5-ZzUSHQs2sgE:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/optbuilder/srfs.go#L178)):*
> ```Go
> 	zip := make(memo.ZipExpr, len(inScope.srfs))
> 	for i, srf := range inScope.srfs {
> 		zip[i] = b.factory.ConstructZipItem(srf.fn, make(opt.ColList, len(srf.cols)))
> ```

Ditto
___
*[pkg/sql/opt/optgen/cmd/optgen/ifactory_gen.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-Lw9YUmp8MGxiYifnJ76:-Lw9YUmq3_fmABPxlfDh:b1k4z46) ([raw file](https://github.com/cockroachdb/cockroach/blob/177b36fdb0a5f078c759f196675f46d7c45c9bcd/pkg/sql/opt/optgen/cmd/optgen/ifactory_gen.go#L66)):*
> ```Go
> 		WithoutTag("Private")
> 
> 	g.w.write("type Factory interface {\n")
> ```

Would be nice to have a comment for Factory (when someone jumps to the definition). It should mention that it is implemented by norm.Factory.
___
*[pkg/sql/opt/props/logical.go, line 342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-Lw9hnZO8leajYHNag9Q:-Lw9hnZO8leajYHNag9R:b-nll7n5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/props/logical.go#L342)):*
> ```Go
> 	// Populated is set to true once the scalar properties have been set by a
> 	// call to a PopulateProps method.
> 	Populated bool
> ```

[nit] This is now only useful for Verify, right? There's a bit of asymmetry here, the Relational props don't have such a field. Though I see the value of it in case direct `Item{}` constructions sneak back in.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Is there any performance downside to populating props eagerly? Or were we pretty much always populating them eventually anyway?

> Reviewed 13 of 13 files at r1, 47 of 47 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-:-LwJihVP2ixeOqMZYu2V:b42fqqc)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andy-kimball and @justinj)

___
*[pkg/sql/opt/memo/check_expr_skip.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-LwJYYMcBAiw3XzS8cq3:-LwJYYMcBAiw3XzS8cq4:bb6ht2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/memo/check_expr_skip.go#L1)):*
> ```Go
> // Copyright 2018 The Cockroach Authors.
> ```

2019
___
*[pkg/sql/opt/optgen/cmd/optgen/exprs_gen.go, line 339 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-LwJdKnp6ozkEziOp1ko:-LwJdKnp6ozkEziOp1kp:bvhb3x7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/optgen/cmd/optgen/exprs_gen.go#L339)):*
> ```Go
> 			fmt.Fprintf(g.w, "func (e *%s) PopulateProps(mem *Memo) {\n", opTyp.name)
> 			fmt.Fprintf(g.w, "  mem.logPropsBuilder.build%sProps(e, &e.scalar)\n", opTyp.name)
> 			fmt.Fprintf(g.w, "  e.scalar.Populated = true\n")
> ```

Is this necessary anymore? If it's not lazy, do we need this `Populated` tag at all?
___
*[pkg/sql/opt/optgen/cmd/optgen/testdata/factory, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-LwJcpm1109WGwe9cx3R:-LwJcpm1109WGwe9cx3S:b-x0jw10) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/optgen/cmd/optgen/testdata/factory#L29)):*
> ```
> }
> 
> [Scalar, Boolean, ListItem]
> ```

Do you want to add the `ScalarProps` tag to test the new call to `PopulateProps`? (Also, shouldn't `Boolean` be `Bool`?)


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43170)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/check_expr.go, line 268 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-Lw9dP0ZEYFmGwP2CsqU:-LwZ1HaK7WnEEXYrvfCC:bpio3b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/edc18a6fc4bb4fbf118f4fc6c98e5642dc6adae7/pkg/sql/opt/memo/check_expr.go#L268)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Out of curiosity, why did this go away? Is this now allowed?
</blockquote></details>

It's still checked above, in `case ScalarPropsExpr:`.
___
*[pkg/sql/opt/memo/check_expr_skip.go, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-LwJYYMcBAiw3XzS8cq3:-LwZ1Q6k6HokekveIhR_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/memo/check_expr_skip.go#L1)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

2019
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/srfs.go, line 116 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-Lw9gO831BGo3zaEAhOY:-LwZ1cFvBX8cOOMFcTDG:boxvgbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/optbuilder/srfs.go#L116)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Not crazy about this, feels strange to call a `Construct` method with a not-yet populated slice (in principle, it could be using those values to populate the props). I'd first populate the slice then call Construct.
</blockquote></details>

Agreed, I was mechanically replacing, and didn't notice. Fixed.
___
*[pkg/sql/opt/optgen/cmd/optgen/exprs_gen.go, line 339 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-LwJdKnp6ozkEziOp1ko:-LwZ5TUD3K19p8uTITIp:b5jpj63) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/optgen/cmd/optgen/exprs_gen.go#L339)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is this necessary anymore? If it's not lazy, do we need this `Populated` tag at all?
</blockquote></details>

`Populated` is used during `checkExpr` to catch bugs where we don't call the factory (and properties are not populated). I did decide to move `Populated` to `Shared` properties, though, so it applies to both Relational and Scalar properties.
___
*[pkg/sql/opt/optgen/cmd/optgen/ifactory_gen.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-Lw9YUmp8MGxiYifnJ76:-LwZ5pY27bKhaz96Ih03:b-2i180w) ([raw file](https://github.com/cockroachdb/cockroach/blob/177b36fdb0a5f078c759f196675f46d7c45c9bcd/pkg/sql/opt/optgen/cmd/optgen/ifactory_gen.go#L66)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Would be nice to have a comment for Factory (when someone jumps to the definition). It should mention that it is implemented by norm.Factory.
</blockquote></details>

The Factory interface is now gone.
___
*[pkg/sql/opt/optgen/cmd/optgen/testdata/factory, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-LwJcpm1109WGwe9cx3R:-LwZ70RJ18hndYHdn8Kx:b-pmck65) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/optgen/cmd/optgen/testdata/factory#L29)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Do you want to add the `ScalarProps` tag to test the new call to `PopulateProps`? (Also, shouldn't `Boolean` be `Bool`?)
</blockquote></details>

Done, also added another case testing lack of `ScalarProps` tag.
___
*[pkg/sql/opt/props/logical.go, line 342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-Lw9hnZO8leajYHNag9Q:-LwZ76PBAYapAFdm4CAf:b-qp2aao) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5a090eadbf52f2d6f44e2235cfa947152455278/pkg/sql/opt/props/logical.go#L342)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] This is now only useful for Verify, right? There's a bit of asymmetry here, the Relational props don't have such a field. Though I see the value of it in case direct `Item{}` constructions sneak back in.
</blockquote></details>

I moved `Populate` to `Shared`, and now check it for all expressions.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 44 of 44 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43170)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @justinj and @RaduBerinde)



<!-- Sent from Reviewable.io -->

andy-kimball commented:
There is some performance impact of this change:
```
name                                    old time/op    new time/op    delta
Phases/kv-read/Parse                      2.90ns ± 4%    2.95ns ±11%    ~     (p=1.000 n=5+5)
Phases/kv-read/OptBuild                   7.57µs ± 2%    7.91µs ± 2%  +4.46%  (p=0.008 n=5+5)
Phases/kv-read/Normalize                  7.60µs ± 5%    7.94µs ± 1%  +4.47%  (p=0.040 n=5+5)
Phases/kv-read/Explore                    13.1µs ± 2%    13.1µs ± 2%    ~     (p=0.690 n=5+5)
Phases/kv-read/ExecBuild                  14.4µs ± 2%    14.7µs ± 2%    ~     (p=0.222 n=5+5)
Phases/kv-read-no-prep/Parse              10.8µs ± 2%    11.1µs ± 2%  +2.57%  (p=0.032 n=5+5)
Phases/kv-read-no-prep/OptBuild           30.8µs ± 2%    31.4µs ± 3%    ~     (p=0.310 n=5+5)
Phases/kv-read-no-prep/Normalize          30.7µs ± 1%    31.6µs ± 3%  +3.11%  (p=0.032 n=5+5)
Phases/kv-read-no-prep/Explore            39.1µs ± 2%    38.7µs ± 1%    ~     (p=0.690 n=5+5)
Phases/kv-read-no-prep/ExecBuild          41.3µs ± 8%    40.9µs ± 3%    ~     (p=1.000 n=5+5)
Phases/kv-read-const/Parse                2.94ns ± 6%    2.84ns ± 3%    ~     (p=0.302 n=5+5)
Phases/kv-read-const/OptBuild              141ns ± 4%     135ns ± 2%  -4.12%  (p=0.016 n=5+5)
Phases/kv-read-const/Normalize             133ns ± 2%     133ns ± 3%    ~     (p=1.000 n=5+5)
Phases/kv-read-const/Explore               145ns ± 2%     148ns ± 3%    ~     (p=0.246 n=5+5)
Phases/kv-read-const/ExecBuild             784ns ± 0%     776ns ± 2%    ~     (p=0.175 n=4+5)
Phases/tpcc-new-order/Parse               3.12ns ± 7%    2.88ns ± 6%  -7.57%  (p=0.032 n=5+5)
Phases/tpcc-new-order/OptBuild            14.2µs ± 5%    15.1µs ± 9%    ~     (p=0.056 n=5+5)
Phases/tpcc-new-order/Normalize           14.3µs ± 3%    14.6µs ± 2%    ~     (p=0.095 n=5+5)
Phases/tpcc-new-order/Explore             41.2µs ± 2%    42.3µs ± 2%    ~     (p=0.095 n=5+5)
Phases/tpcc-new-order/ExecBuild           45.1µs ± 5%    44.8µs ± 2%    ~     (p=0.548 n=5+5)
Phases/tpcc-delivery/Parse                2.96ns ± 7%    2.88ns ± 8%    ~     (p=0.389 n=5+5)
Phases/tpcc-delivery/OptBuild             12.6µs ± 2%    13.2µs ± 2%  +4.65%  (p=0.008 n=5+5)
Phases/tpcc-delivery/Normalize            12.8µs ± 3%    13.4µs ± 3%  +4.21%  (p=0.008 n=5+5)
Phases/tpcc-delivery/Explore              29.5µs ± 2%    30.0µs ± 2%    ~     (p=0.222 n=5+5)
Phases/tpcc-delivery/ExecBuild            32.0µs ± 2%    31.9µs ± 2%    ~     (p=0.548 n=5+5)
Phases/tpcc-stock-level/Parse             2.90ns ± 6%    3.18ns ±18%    ~     (p=0.198 n=5+5)
Phases/tpcc-stock-level/OptBuild          54.0µs ± 1%    55.0µs ± 1%  +1.86%  (p=0.032 n=5+5)
Phases/tpcc-stock-level/Normalize         54.7µs ± 2%    55.5µs ± 2%    ~     (p=0.151 n=5+5)
Phases/tpcc-stock-level/Explore            195µs ± 1%     197µs ± 1%    ~     (p=0.095 n=5+5)
Phases/tpcc-stock-level/ExecBuild          205µs ± 1%     206µs ± 2%    ~     (p=0.548 n=5+5)
EndToEnd/kv-read/EndToEnd                  179µs ± 3%     181µs ± 3%    ~     (p=0.421 n=5+5)
EndToEnd/kv-read-no-prep/EndToEnd          278µs ± 4%     275µs ± 1%    ~     (p=1.000 n=5+5)
EndToEnd/kv-read-const/EndToEnd            148µs ± 2%     152µs ± 1%  +2.49%  (p=0.032 n=5+5)
EndToEnd/tpcc-new-order/EndToEnd           239µs ± 1%     242µs ± 3%    ~     (p=0.310 n=5+5)
EndToEnd/tpcc-delivery/EndToEnd            210µs ± 3%     211µs ± 1%    ~     (p=0.310 n=5+5)
EndToEnd/tpcc-stock-level/EndToEnd         540µs ± 2%     559µs ± 5%    ~     (p=0.151 n=5+5)
Chain/chain-1/Explore                     15.6µs ± 3%    16.3µs ± 8%    ~     (p=0.151 n=5+5)
Chain/chain-2/Explore                     79.0µs ± 3%    79.6µs ± 2%    ~     (p=0.222 n=5+5)
Chain/chain-3/Explore                      141µs ± 1%     149µs ± 1%  +5.48%  (p=0.008 n=5+5)
Chain/chain-4/Explore                      202µs ± 1%     214µs ± 1%  +5.84%  (p=0.008 n=5+5)
Chain/chain-5/Explore                      272µs ± 2%     285µs ± 2%  +4.90%  (p=0.008 n=5+5)
Chain/chain-6/Explore                      344µs ± 1%     361µs ± 1%  +5.07%  (p=0.008 n=5+5)
Chain/chain-7/Explore                      414µs ± 1%     434µs ± 1%  +4.89%  (p=0.008 n=5+5)
Chain/chain-8/Explore                      492µs ± 2%     517µs ± 1%  +5.03%  (p=0.008 n=5+5)
Chain/chain-9/Explore                      632µs ± 7%     642µs ± 2%    ~     (p=0.421 n=5+5)
Chain/chain-10/Explore                     710µs ± 3%     741µs ± 2%  +4.39%  (p=0.016 n=5+5)
Chain/chain-11/Explore                     824µs ± 5%     839µs ± 1%    ~     (p=0.421 n=5+5)
Chain/chain-12/Explore                     919µs ± 2%     986µs ± 8%  +7.30%  (p=0.008 n=5+5)
Chain/chain-13/Explore                    1.06ms ± 7%    1.09ms ± 1%    ~     (p=0.151 n=5+5)
Chain/chain-14/Explore                    1.23ms ± 2%    1.28ms ± 1%  +4.10%  (p=0.008 n=5+5)
Chain/chain-15/Explore                    1.38ms ± 2%    1.43ms ± 1%  +3.37%  (p=0.008 n=5+5)
Chain/chain-16/Explore                    1.54ms ± 2%    1.59ms ± 2%  +3.11%  (p=0.032 n=5+5)
Chain/chain-17/Explore                    1.70ms ± 2%    1.75ms ± 3%  +3.40%  (p=0.008 n=5+5)
Chain/chain-18/Explore                    1.87ms ± 3%    1.95ms ± 2%  +4.04%  (p=0.016 n=5+5)
Chain/chain-19/Explore                    2.03ms ± 2%    2.12ms ± 3%  +4.34%  (p=0.008 n=5+5)
FKInsert/SingleRow/None                    672µs ± 3%     673µs ± 1%    ~     (p=0.548 n=5+5)
FKInsert/SingleRow/Old                     726µs ± 2%     733µs ± 2%    ~     (p=0.548 n=5+5)
FKInsert/SingleRow/New                    1.11ms ± 5%    1.08ms ± 2%    ~     (p=0.151 n=5+5)
FKInsert/SingleRow/FastPath                779µs ± 1%     773µs ± 2%    ~     (p=0.310 n=5+5)
FKInsert/MultiRowSingleParent/None        1.38ms ± 1%    1.36ms ± 2%    ~     (p=0.056 n=5+5)
FKInsert/MultiRowSingleParent/Old         2.07ms ± 4%    2.06ms ± 3%    ~     (p=1.000 n=5+5)
FKInsert/MultiRowSingleParent/New         2.02ms ± 4%    2.10ms ± 7%    ~     (p=0.056 n=5+5)
FKInsert/MultiRowSingleParent/FastPath    1.47ms ± 2%    1.54ms ±12%    ~     (p=0.222 n=5+5)
FKInsert/MultiRowMultiParent/None         1.43ms ± 2%    1.44ms ± 2%    ~     (p=0.548 n=5+5)
FKInsert/MultiRowMultiParent/Old          2.35ms ± 7%    2.19ms ± 2%  -6.65%  (p=0.032 n=5+5)
FKInsert/MultiRowMultiParent/New          2.36ms ± 4%    2.30ms ± 2%    ~     (p=0.056 n=5+5)
FKInsert/MultiRowMultiParent/FastPath     1.74ms ± 7%    1.71ms ± 1%    ~     (p=0.841 n=5+5)

name                                    old alloc/op   new alloc/op   delta
Phases/kv-read/Parse                       0.00B          0.00B         ~     (all equal)
Phases/kv-read/OptBuild                   2.86kB ± 0%    2.86kB ± 0%    ~     (all equal)
Phases/kv-read/Normalize                  2.86kB ± 0%    2.86kB ± 0%    ~     (all equal)
Phases/kv-read/Explore                    4.34kB ± 0%    4.34kB ± 0%    ~     (all equal)
Phases/kv-read/ExecBuild                  4.56kB ± 0%    4.56kB ± 0%    ~     (all equal)
Phases/kv-read-no-prep/Parse              3.58kB ± 0%    3.58kB ± 0%    ~     (all equal)
Phases/kv-read-no-prep/OptBuild           8.98kB ± 0%    8.98kB ± 0%    ~     (all equal)
Phases/kv-read-no-prep/Normalize          8.98kB ± 0%    8.98kB ± 0%    ~     (all equal)
Phases/kv-read-no-prep/Explore            10.5kB ± 0%    10.5kB ± 0%    ~     (all equal)
Phases/kv-read-no-prep/ExecBuild          10.7kB ± 0%    10.7kB ± 0%    ~     (all equal)
Phases/kv-read-const/Parse                 0.00B          0.00B         ~     (all equal)
Phases/kv-read-const/OptBuild              48.0B ± 0%     48.0B ± 0%    ~     (all equal)
Phases/kv-read-const/Normalize             48.0B ± 0%     48.0B ± 0%    ~     (all equal)
Phases/kv-read-const/Explore               48.0B ± 0%     48.0B ± 0%    ~     (all equal)
Phases/kv-read-const/ExecBuild              264B ± 0%      264B ± 0%    ~     (all equal)
Phases/tpcc-new-order/Parse                0.00B          0.00B         ~     (all equal)
Phases/tpcc-new-order/OptBuild            4.73kB ± 0%    4.70kB ± 0%  -0.68%  (p=0.008 n=5+5)
Phases/tpcc-new-order/Normalize           4.73kB ± 0%    4.70kB ± 0%  -0.68%  (p=0.008 n=5+5)
Phases/tpcc-new-order/Explore             11.1kB ± 0%    11.1kB ± 0%  -0.58%  (p=0.008 n=5+5)
Phases/tpcc-new-order/ExecBuild           11.4kB ± 0%    11.3kB ± 0%  -0.56%  (p=0.000 n=5+4)
Phases/tpcc-delivery/Parse                 0.00B          0.00B         ~     (all equal)
Phases/tpcc-delivery/OptBuild             4.70kB ± 0%    4.70kB ± 0%    ~     (all equal)
Phases/tpcc-delivery/Normalize            4.73kB ± 0%    4.73kB ± 0%    ~     (all equal)
Phases/tpcc-delivery/Explore              8.65kB ± 0%    8.65kB ± 0%    ~     (all equal)
Phases/tpcc-delivery/ExecBuild            8.85kB ± 0%    8.85kB ± 0%    ~     (p=0.881 n=5+5)
Phases/tpcc-stock-level/Parse              0.00B          0.00B         ~     (all equal)
Phases/tpcc-stock-level/OptBuild          15.8kB ± 0%    15.8kB ± 0%  +0.41%  (p=0.008 n=5+5)
Phases/tpcc-stock-level/Normalize         15.4kB ± 0%    15.4kB ± 0%  +0.21%  (p=0.016 n=4+5)
Phases/tpcc-stock-level/Explore           50.4kB ± 0%    51.2kB ± 0%  +1.58%  (p=0.008 n=5+5)
Phases/tpcc-stock-level/ExecBuild         51.3kB ± 0%    52.1kB ± 0%  +1.56%  (p=0.008 n=5+5)
EndToEnd/kv-read/EndToEnd                 23.2kB ± 0%    23.2kB ± 0%    ~     (p=1.000 n=5+5)
EndToEnd/kv-read-no-prep/EndToEnd         28.9kB ± 0%    28.9kB ± 0%    ~     (p=0.686 n=4+4)
EndToEnd/kv-read-const/EndToEnd           18.7kB ± 0%    18.7kB ± 0%    ~     (p=0.886 n=4+4)
EndToEnd/tpcc-new-order/EndToEnd          46.5kB ± 0%    46.4kB ± 0%  -0.21%  (p=0.029 n=4+4)
EndToEnd/tpcc-delivery/EndToEnd           29.7kB ± 0%    29.7kB ± 0%    ~     (p=0.952 n=5+4)
EndToEnd/tpcc-stock-level/EndToEnd         139kB ± 0%     140kB ± 0%  +0.59%  (p=0.016 n=4+5)
Chain/chain-1/Explore                     6.38kB ± 0%    6.38kB ± 0%    ~     (all equal)
Chain/chain-2/Explore                     21.9kB ± 0%    22.0kB ± 0%  +0.59%  (p=0.008 n=5+5)
Chain/chain-3/Explore                     38.1kB ± 0%    40.4kB ± 0%  +6.13%  (p=0.008 n=5+5)
Chain/chain-4/Explore                     56.2kB ± 0%    60.1kB ± 0%  +7.06%  (p=0.008 n=5+5)
Chain/chain-5/Explore                     77.4kB ± 0%    83.6kB ± 0%  +7.94%  (p=0.008 n=5+5)
Chain/chain-6/Explore                      103kB ± 0%     111kB ± 0%  +8.03%  (p=0.008 n=5+5)
Chain/chain-7/Explore                      127kB ± 0%     138kB ± 0%  +8.79%  (p=0.008 n=5+5)
Chain/chain-8/Explore                      150kB ± 0%     164kB ± 0%  +9.44%  (p=0.008 n=5+5)
Chain/chain-9/Explore                      203kB ± 0%     220kB ± 0%  +8.61%  (p=0.008 n=5+5)
Chain/chain-10/Explore                     235kB ± 0%     255kB ± 0%  +8.79%  (p=0.008 n=5+5)
Chain/chain-11/Explore                     267kB ± 0%     291kB ± 0%  +9.11%  (p=0.008 n=5+5)
Chain/chain-12/Explore                     315kB ± 0%     343kB ± 0%  +8.76%  (p=0.008 n=5+5)
Chain/chain-13/Explore                     356kB ± 0%     389kB ± 0%  +9.26%  (p=0.008 n=5+5)
Chain/chain-14/Explore                     457kB ± 0%     494kB ± 0%  +8.09%  (p=0.008 n=5+5)
Chain/chain-15/Explore                     505kB ± 0%     549kB ± 0%  +8.54%  (p=0.008 n=5+5)
Chain/chain-16/Explore                     581kB ± 0%     629kB ± 0%  +8.26%  (p=0.008 n=5+5)
Chain/chain-17/Explore                     640kB ± 0%     694kB ± 0%  +8.45%  (p=0.008 n=5+5)
Chain/chain-18/Explore                     704kB ± 0%     764kB ± 0%  +8.59%  (p=0.008 n=5+5)
Chain/chain-19/Explore                     768kB ± 0%     835kB ± 0%  +8.70%  (p=0.008 n=5+5)
FKInsert/SingleRow/None                   55.2kB ± 0%    55.2kB ± 0%    ~     (p=0.222 n=5+5)
FKInsert/SingleRow/Old                    62.4kB ± 0%    62.4kB ± 0%    ~     (p=0.460 n=5+5)
FKInsert/SingleRow/New                     124kB ± 0%     124kB ± 0%  +0.31%  (p=0.016 n=4+5)
FKInsert/SingleRow/FastPath               67.4kB ± 0%    67.6kB ± 0%  +0.29%  (p=0.008 n=5+5)
FKInsert/MultiRowSingleParent/None         210kB ± 0%     210kB ± 0%    ~     (p=0.595 n=5+5)
FKInsert/MultiRowSingleParent/Old          301kB ± 0%     301kB ± 0%    ~     (p=1.000 n=5+5)
FKInsert/MultiRowSingleParent/New          325kB ± 0%     325kB ± 0%    ~     (p=0.056 n=5+5)
FKInsert/MultiRowSingleParent/FastPath     223kB ± 0%     223kB ± 0%    ~     (p=0.056 n=5+5)
FKInsert/MultiRowMultiParent/None          213kB ± 0%     213kB ± 0%    ~     (p=0.548 n=5+5)
FKInsert/MultiRowMultiParent/Old           307kB ± 0%     307kB ± 0%    ~     (p=0.548 n=5+5)
FKInsert/MultiRowMultiParent/New           354kB ± 0%     354kB ± 0%  +0.08%  (p=0.008 n=5+5)
FKInsert/MultiRowMultiParent/FastPath      241kB ± 0%     241kB ± 0%    ~     (p=0.151 n=5+5)

name                                    old allocs/op  new allocs/op  delta
Phases/kv-read/Parse                        0.00           0.00         ~     (all equal)
Phases/kv-read/OptBuild                     24.0 ± 0%      24.0 ± 0%    ~     (all equal)
Phases/kv-read/Normalize                    24.0 ± 0%      24.0 ± 0%    ~     (all equal)
Phases/kv-read/Explore                      32.0 ± 0%      32.0 ± 0%    ~     (all equal)
Phases/kv-read/ExecBuild                    35.0 ± 0%      35.0 ± 0%    ~     (all equal)
Phases/kv-read-no-prep/Parse                29.0 ± 0%      29.0 ± 0%    ~     (all equal)
Phases/kv-read-no-prep/OptBuild             70.0 ± 0%      70.0 ± 0%    ~     (all equal)
Phases/kv-read-no-prep/Normalize            70.0 ± 0%      70.0 ± 0%    ~     (all equal)
Phases/kv-read-no-prep/Explore              78.0 ± 0%      78.0 ± 0%    ~     (all equal)
Phases/kv-read-no-prep/ExecBuild            81.0 ± 0%      81.0 ± 0%    ~     (all equal)
Phases/kv-read-const/Parse                  0.00           0.00         ~     (all equal)
Phases/kv-read-const/OptBuild               1.00 ± 0%      1.00 ± 0%    ~     (all equal)
Phases/kv-read-const/Normalize              1.00 ± 0%      1.00 ± 0%    ~     (all equal)
Phases/kv-read-const/Explore                1.00 ± 0%      1.00 ± 0%    ~     (all equal)
Phases/kv-read-const/ExecBuild              4.00 ± 0%      4.00 ± 0%    ~     (all equal)
Phases/tpcc-new-order/Parse                 0.00           0.00         ~     (all equal)
Phases/tpcc-new-order/OptBuild              30.0 ± 0%      30.0 ± 0%    ~     (all equal)
Phases/tpcc-new-order/Normalize             30.0 ± 0%      30.0 ± 0%    ~     (all equal)
Phases/tpcc-new-order/Explore               62.0 ± 0%      62.0 ± 0%    ~     (all equal)
Phases/tpcc-new-order/ExecBuild             67.0 ± 0%      67.0 ± 0%    ~     (all equal)
Phases/tpcc-delivery/Parse                  0.00           0.00         ~     (all equal)
Phases/tpcc-delivery/OptBuild               25.0 ± 0%      25.0 ± 0%    ~     (all equal)
Phases/tpcc-delivery/Normalize              26.0 ± 0%      26.0 ± 0%    ~     (all equal)
Phases/tpcc-delivery/Explore                49.0 ± 0%      49.0 ± 0%    ~     (all equal)
Phases/tpcc-delivery/ExecBuild              53.0 ± 0%      53.0 ± 0%    ~     (all equal)
Phases/tpcc-stock-level/Parse               0.00           0.00         ~     (all equal)
Phases/tpcc-stock-level/OptBuild            98.0 ± 0%      98.0 ± 0%    ~     (all equal)
Phases/tpcc-stock-level/Normalize           96.0 ± 0%      96.0 ± 0%    ~     (all equal)
Phases/tpcc-stock-level/Explore              272 ± 0%       272 ± 0%    ~     (all equal)
Phases/tpcc-stock-level/ExecBuild            284 ± 0%       284 ± 0%    ~     (all equal)
EndToEnd/kv-read/EndToEnd                    231 ± 0%       231 ± 0%    ~     (p=0.333 n=5+4)
EndToEnd/kv-read-no-prep/EndToEnd            297 ± 0%       297 ± 0%    ~     (all equal)
EndToEnd/kv-read-const/EndToEnd              193 ± 0%       193 ± 0%    ~     (all equal)
EndToEnd/tpcc-new-order/EndToEnd             290 ± 0%       290 ± 0%    ~     (all equal)
EndToEnd/tpcc-delivery/EndToEnd              265 ± 0%       265 ± 0%    ~     (all equal)
EndToEnd/tpcc-stock-level/EndToEnd           741 ± 0%       741 ± 0%    ~     (all equal)
Chain/chain-1/Explore                       40.0 ± 0%      40.0 ± 0%    ~     (all equal)
Chain/chain-2/Explore                        156 ± 0%       156 ± 0%    ~     (all equal)
Chain/chain-3/Explore                        280 ± 0%       290 ± 0%  +3.57%  (p=0.008 n=5+5)
Chain/chain-4/Explore                        398 ± 0%       413 ± 0%  +3.87%  (p=0.016 n=4+5)
Chain/chain-5/Explore                        520 ± 0%       542 ± 0%  +4.23%  (p=0.008 n=5+5)
Chain/chain-6/Explore                        644 ± 0%       672 ± 0%  +4.35%  (p=0.008 n=5+5)
Chain/chain-7/Explore                        770 ± 0%       804 ± 0%  +4.42%  (p=0.008 n=5+5)
Chain/chain-8/Explore                        898 ± 0%       938 ± 0%  +4.45%  (p=0.008 n=5+5)
Chain/chain-9/Explore                      1.05k ± 0%     1.09k ± 0%  +4.40%  (p=0.008 n=5+5)
Chain/chain-10/Explore                     1.18k ± 0%     1.23k ± 0%  +4.41%  (p=0.008 n=5+5)
Chain/chain-11/Explore                     1.32k ± 0%     1.37k ± 0%  +4.41%  (p=0.016 n=4+5)
Chain/chain-12/Explore                     1.46k ± 0%     1.52k ± 0%  +4.40%  (p=0.008 n=5+5)
Chain/chain-13/Explore                     1.60k ± 0%     1.67k ± 0%  +4.38%  (p=0.016 n=4+5)
Chain/chain-14/Explore                     1.77k ± 0%     1.84k ± 0%  +4.30%  (p=0.029 n=4+4)
Chain/chain-15/Explore                     1.92k ± 0%     2.00k ± 0%  +4.28%  (p=0.016 n=5+4)
Chain/chain-16/Explore                     2.07k ± 0%     2.16k ± 0%  +4.27%  (p=0.016 n=5+4)
Chain/chain-17/Explore                     2.22k ± 0%     2.32k ± 0%  +4.23%  (p=0.016 n=5+4)
Chain/chain-18/Explore                     2.38k ± 0%     2.48k ± 0%  +4.20%  (p=0.008 n=5+5)
Chain/chain-19/Explore                     2.54k ± 0%     2.64k ± 0%  +4.16%  (p=0.016 n=4+5)
FKInsert/SingleRow/None                      557 ± 0%       557 ± 0%    ~     (p=0.444 n=5+5)
FKInsert/SingleRow/Old                       643 ± 0%       644 ± 0%    ~     (p=1.000 n=5+5)
FKInsert/SingleRow/New                     1.07k ± 0%     1.07k ± 0%    ~     (p=0.683 n=5+5)
FKInsert/SingleRow/FastPath                  666 ± 0%       666 ± 0%    ~     (p=0.444 n=5+5)
FKInsert/MultiRowSingleParent/None         1.63k ± 0%     1.63k ± 0%    ~     (all equal)
FKInsert/MultiRowSingleParent/Old          2.69k ± 0%     2.69k ± 0%    ~     (p=1.000 n=5+5)
FKInsert/MultiRowSingleParent/New          2.62k ± 0%     2.62k ± 0%    ~     (p=1.000 n=4+5)
FKInsert/MultiRowSingleParent/FastPath     1.78k ± 0%     1.78k ± 0%    ~     (p=0.333 n=4+5)
FKInsert/MultiRowMultiParent/None          1.61k ± 0%     1.61k ± 0%    ~     (p=1.000 n=5+5)
FKInsert/MultiRowMultiParent/Old           2.65k ± 0%     2.65k ± 0%    ~     (p=0.079 n=5+5)
FKInsert/MultiRowMultiParent/New           2.73k ± 0%     2.73k ± 0%    ~     (p=0.929 n=5+5)
FKInsert/MultiRowMultiParent/FastPath      1.88k ± 0%     1.88k ± 0%    ~     (p=1.000 n=5+4)
```

The `Chain` diffs are due to the `SimplifyFilters` rule, where we transform filters like `[(And $left $right)]` into `[$left, $right]`. In the old code, we constructed the scalar props lazily, and so didn't construct the intermediate props for the first filter list. In the new code, we construct props for both the first and second list.

I don't see a good way to avoid this. But I think the benefits of this change outweigh this problem, so I think we should go ahead anyway.

andy-kimball committed:
opt: do not lazily construct ScalarProps

Previously, ListItem scalar operators had a method like this:

  ScalarProps(mem *memo.Memo)

It lazily constructed scalar properties when called. The problem was
that this required the Memo to be passed to many contexts simply for
the purpose of calling ScalarProps, which was inconvenient. In
addition, there will be thread-safety issues if we ever call this
method after the expression tree becomes immutable (e.g. after it's
been added to the query cache).

To fix this, this patch constructs the ScalarProps greedily rather
than lazily. All locations that directly constructed ListItem structs
now call into new Factory methods that both construct the structs and
populate them with scalar properties. This way, the scalar properties
are immutable and always available.

Fixes #36148

Release note: none

RaduBerinde reviewed (approved):
I agree, <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 16 of 47 files at r2, 30 of 44 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43170#-:-LwcVreT9POq6wy8S2NL:bp7m0fk)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @justinj)



<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+
